File: API/Service/AuthService/AuthService.cs
using API.Common;
using API.Dtos.Auth;
using API.Dtos.User;
using API.Entities;
using API.Interfaces;
using API.UnitOfWork;
using Microsoft.AspNetCore.Identity;

namespace API.Service.AuthService
{
    public class AuthService
    {
        private readonly TokenService _tokenService;
        private readonly IUnitOfWork _uow;
        private readonly IAuthRepository _authRepository;
        private readonly IPasswordHasher<User> _passwordHasher;

        public AuthService(
            TokenService tokenService,
            IUnitOfWork uow,
            IAuthRepository authRepository,
            IPasswordHasher<User> passwordHasher)
        {
            _tokenService = tokenService;
            _uow = uow;
            _authRepository = authRepository;
            _passwordHasher = passwordHasher;
        }

        // LOGIN
        public async Task<Result<LoginResponseDto>> LoginHandleAsync(LoginRequestDto model)
        {
            var user = await _authRepository.GetByUserNameAsync(model.UserName);

            if (user == null)
                return Result<LoginResponseDto>.Failure("User not found");

            var verify = _passwordHasher.VerifyHashedPassword(user, user.PasswordHash, model.Password);

            if (verify == PasswordVerificationResult.Failed)
                return Result<LoginResponseDto>.Failure("Invalid username or password");

            var tokenDto = GenerateTokens(user);

            await _uow.SaveChangesAsync();

            return Result<LoginResponseDto>.SuccessResult(tokenDto, "Login successful");
        }

        // REFRESH TOKEN
        public async Task<Result<LoginResponseDto>> RefreshTokenAsync(string refreshToken)
        {
            var user = await _authRepository.GetByRefreshTokenAsync(refreshToken);

            if (user == null)
                return Result<LoginResponseDto>.Failure("Invalid refresh token");

            if (user.RefreshTokenExpiryTime == null || user.RefreshTokenExpiryTime < DateTime.UtcNow)
                return Result<LoginResponseDto>.Failure("Refresh token expired");

            var tokenDto = GenerateTokens(user);

            await _uow.SaveChangesAsync();

            return Result<LoginResponseDto>.SuccessResult(tokenDto, "Refresh successful");
        }

        //  LOGOUT
        public async Task<Result<bool>> LogoutAsync(int userId)
        {
            var user = await _authRepository.GetByIdAsync(userId);

            if (user == null)
                return Result<bool>.Failure("User not found");

            user.RefreshToken = null;
            user.RefreshTokenExpiryTime = null;

            await _uow.SaveChangesAsync();

            return Result<bool>.SuccessResult(true, "Logout successful");
        }

        // REGISTER
        public async Task<Result<RegisterResponseDto>> RegisterHandleAsync(RegisterRequestDto model)
        {
            var existingUser = await _authRepository.GetByUserNameAsync(model.Username);

            if (existingUser != null)
                return Result<RegisterResponseDto>.Failure("Username already exists");

            var user = new User
            {
                UserName = model.Username,
                Email = model.Email,
                CreatedAt = DateTime.UtcNow
            };

            user.PasswordHash = _passwordHasher.HashPassword(user, model.Password);

            await _authRepository.AddAsync(user);
            await _uow.SaveChangesAsync();

            return Result<RegisterResponseDto>.SuccessResult(
                new RegisterResponseDto
                {
                    Username = user.UserName,
                    Email = user.Email
                },
                "User registered successfully"
            );
        }

        // GET ME
        public async Task<UserDto> GetMeAsync(int userId)
        {
            var user = await _authRepository.GetByIdAsync(userId);

            if (user == null)
                throw new Exception("User not found");

            return new UserDto
            {
                Username = user.UserName,
                Email = user.Email
            };
        }

        // GENERATE TOKENS
        private LoginResponseDto GenerateTokens(User user)
        {
            var (accessToken, expires) = _tokenService.GenerateToken(user);
            var refreshToken = _tokenService.GenerateRefreshToken();

            user.RefreshToken = refreshToken;
            user.RefreshTokenExpiryTime = DateTime.UtcNow.AddDays(7);
            return new LoginResponseDto
            {
                Username = user.UserName,
                AccessToken = accessToken,
                RefreshToken = refreshToken,
                AccessTokenExpiry = expires
            };
        }
    }
}

File: Frontend/src/services/authService.js
import axiosClient from '../api/axiosClient';

const authService = {
  /**
   * Đăng nhập
   * POST /api/auth/login
   * Body: { userName, password }
   * Response: { username, token }
   */
  login: async (userName, password) => {
    const response = await axiosClient.post('/auth/login', { userName, password });
    return response;
  },

  /**
   * Refresh Token
   * POST /api/auth/refresh
   * Body: { refreshToken }
   */
  refresh: async (refreshToken) => {
    const response = await axiosClient.post('/auth/refresh', { refreshToken });
    return response;
  },

  /**
   * Đăng ký
   * POST /api/auth/register
   * Body: { username, email, password, confirmPassword }
   */
  register: async (username, email, password, confirmPassword) => {
    const response = await axiosClient.post('/auth/register', {
      username,
      email,
      password,
      confirmPassword,
    });
    return response;
  },

  /**
   * Lấy thông tin user hiện tại
   * GET /api/auth/me  (yêu cầu Bearer token)
   */
  getMe: async () => {
    const response = await axiosClient.get('/auth/me');
    return response;
  },

  /** Lưu token và username vào localStorage */
  saveSession: (accessToken, refreshToken, username) => {
    localStorage.setItem('access_token', accessToken);
    localStorage.setItem('refresh_token', refreshToken);
    localStorage.setItem('username', username);
  },

  /** Xoá phiên đăng nhập */
  logout: () => {
    localStorage.removeItem('access_token');
    localStorage.removeItem('refresh_token');
    localStorage.removeItem('username');
  },

  /** Kiểm tra đã đăng nhập chưa */
  isAuthenticated: () => {
    return !!localStorage.getItem('access_token');
  },

  /** Lấy username đang đăng nhập */
  getUsername: () => {
    return localStorage.getItem('username') || '';
  },
};

export default authService;
